lib/SPIRV/SPIRVToOCL12.cpp

```diff
@@ -114,6 +138,91 @@ void SPIRVToOCL12::visitCallSPIRVControlBarrier(CallInst *CI) {
       &Attrs);
 }
 
+Instruction *SPIRVToOCL12::visitCallSPIRVAtomicIncDec(CallInst *CI, Op OC) {
+  AttributeList Attrs = CI->getCalledFunction()->getAttributes();
+  return mutateCallInstOCL(
+      M, CI,
+      [=](CallInst *, std::vector<Value *> &Args) {
+        Args.resize(1);
+        return OCLSPIRVBuiltinMap::rmap(OC);
+      },
+      &Attrs);
+}
+
+CallInst *SPIRVToOCL12::mutateCommonAtomicArguments(CallInst *CI, Op OC) {
+  assert(CI->getCalledFunction() && "Unexpected indirect call");
+  AttributeList Attrs = CI->getCalledFunction()->getAttributes();
+
+  return mutateCallInstOCL(
+      M, CI,
+      [=](CallInst *, std::vector<Value *> &Args) {
+        auto Ptr = findFirstPtr(Args);
+        auto NumOrder = getSPIRVAtomicBuiltinNumMemoryOrderArgs(OC);
+        auto ArgsToRemove = NumOrder + 1; // OpenCL1.2 builtins does not use
+                                          // scope and memory order arguments
+        auto StartIdx = Ptr + 1;
+        auto StopIdx = StartIdx + ArgsToRemove;
+        Args.erase(Args.begin() + StartIdx, Args.begin() + StopIdx);
+        return CI->getName();
+      },
+      &Attrs);
+}
+
+Instruction *SPIRVToOCL12::visitCallSPIRVAtomicUMinUMax(CallInst *CI, Op OC) {
+  AttributeList Attrs = CI->getCalledFunction()->getAttributes();
+  return mutateCallInstOCL(
+      M, CI,
+      [=](CallInst *, std::vector<Value *> &Args) {
+        return OCL12SPIRVBuiltinMap::rmap(OC == OpAtomicUMin ? OpAtomicSMin
+                                                             : OpAtomicSMax);
+      },
+      &Attrs);
+}
+
+Instruction *SPIRVToOCL12::visitCallSPIRVAtomicCmpExchg(CallInst *CI, Op OC) {
+  AttributeList Attrs = CI->getCalledFunction()->getAttributes();
+  return mutateCallInstOCL(
+      M, CI,
+      [=](CallInst *, std::vector<Value *> &Args) {
+        // SPIRV OpAtomicCompareExchange and OpAtomicCompareExchangeWeak
+        // has Value and Comparator in different order than ocl functions
+        // both of them are translated into atomic_cmpxchg
+        std::swap(Args[1], Args[2]);
+        return OCL12SPIRVBuiltinMap::rmap(OpAtomicCompareExchange);
+      },
+      &Attrs);
+}
+
+Instruction *SPIRVToOCL12::visitCallSPIRVAtomicBuiltin(CallInst *CI, Op OC) {
+  CallInst *CIG = mutateCommonAtomicArguments(CI, OC);
+
+  Instruction *NewCI = nullptr;
+  switch (OC) {
+  case OpAtomicUMin:
+  case OpAtomicUMax:
```

Comment:
What about OpAtomic**S**Min and OpAtomic**S**Max ?

Reply:
Those opcodes are handled directly using map here: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/263/files#diff-ac841da0a76051bb8d965e112c5b3836R708

OpAtomicUMin and OpAtomicUMax are handle here, because those are translated to the same builtins in llvm and couldn't be added to the map.